tests/unit/test_charm.py

```diff
@@ -1,15 +1,160 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-import unittest
+import json
+from typing import Tuple
 
+import yaml
+from ops.model import ActiveStatus, WaitingStatus
 from ops.testing import Harness
 
-from charm import OathkeeperCharm
+CONTAINER_NAME = "oathkeeper"
 
 
-class TestCharm(unittest.TestCase):
-    def setUp(self):
-        self.harness = Harness(OathkeeperCharm)
-        self.addCleanup(self.harness.cleanup)
-        self.harness.begin()
+def setup_ingress_relation(harness: Harness) -> Tuple[int, str]:
+    """Set up ingress relation."""
+    harness.set_leader(True)
+    relation_id = harness.add_relation("ingress", "traefik")
+    harness.add_relation_unit(relation_id, "traefik/0")
+    url = f"http://ingress:80/{harness.model.name}-oathkeeper"
+    harness.update_relation_data(
+        relation_id,
+        "traefik",
+        {"ingress": json.dumps({"url": url})},
+    )
+    return relation_id, url
+
+
+def test_pebble_container_can_connect(harness: Harness) -> None:
+    harness.set_can_connect(CONTAINER_NAME, True)
+    harness.charm.on.oathkeeper_pebble_ready.emit(CONTAINER_NAME)
+
+    assert isinstance(harness.charm.unit.status, ActiveStatus)
+    service = harness.model.unit.get_container(CONTAINER_NAME).get_service("oathkeeper")
+    assert service.is_running()
+
+
+def test_pebble_container_cannot_connect(harness: Harness) -> None:
+    harness.set_can_connect(CONTAINER_NAME, False)
+    harness.charm.on.oathkeeper_pebble_ready.emit(CONTAINER_NAME)
+
+    assert harness.charm.unit.status == WaitingStatus("Waiting to connect to Oathkeeper container")
+
+
+def test_update_container_config(harness: Harness) -> None:
+    harness.set_can_connect(CONTAINER_NAME, True)
+
+    harness.charm.on.oathkeeper_pebble_ready.emit(CONTAINER_NAME)
+
+    expected_config = {
```

Comment:
Have a tiny question about the necessity to construct the expected Oathkeeper configuration in this way. In this test case, I think what we would like to make sure is that the charm actually reads the Jinja template file and renders the configuration file with the correct parameters.

The inputs to the method `_render_conf_file` actually would be the template file and the two Kratos' parameters. So is it better that we use the expected template file and expected Kratos' parameters to render the configuration file as the `expected_config` instead of pasting the whole rendered configuration?

Reply:
It's a common pattern in our tests to check for the full rendered configuration file. In (near) future there will be more parameters.
That said, I will modify this test to check the `expected_config` against what was actually pushed to the container (use `container_config.read()`) to not rely on the `_render_conf_file` method.